samples: Bluetooth: Added multiple broadcaster and Extended Advertising chain PDU validation

cvinayak opened the pull request:
Added Extended Advertising multiple advertising set
broadcaster sample. Which is used to manually verify that
AD data PDU chaining is functional.

Added implementation to Observer sample so that when
Extended Advertising feature is enabled, it can receive
Extended Advertising Reports, and hence receive long AD
data from peer Extended Advertising Broadcaster Multiple
sample.

Fix Extended Advertising and Periodic Advertising's parent
PDU's aux ptr field to contain correct aux offset value to
its chain PDU.

Uninitialized pointer reference to aux ptr in the parent PDU
was used before the pointer reference was returned by the
function adding the aux ptr fields in the parent PDU.

Use local variable to update current PDU's Common Extended
Header Format flags instead of updating it directly in the
PDU buffer. This is required to be able to non-destructively
be able to manipulate the PDU's header flags and contents
many time before a final commit of the updated PDU.

In order to calculate overflowed AD data length, PDU is
manipulated couple of times to determine the overflow length
and next to be able to add aux ptr field when chain PDU is
appended.

The changes are now consistent with how Periodic Advertising
PDU's Common Extended Header Format is manipulated.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
samples: Bluetooth: Multiple Broadcaster

Added Extended Advertising multiple advertising set
broadcaster sample.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Use local variable for current PDU's flags

Use local variable to update current PDU's Common Extended
Header Format flags instead of updating it directly in the
PDU buffer. This is required to be able to non-destructively
be able to manipulate the PDU's header flags and contents
many time before a final commit of the updated PDU.

In order to calculate overflowed AD data length, PDU is
manipulated couple of times to determine the overflow length
and next to be able to add aux ptr field when chain PDU is
appended.

The changes are now consistent with how Periodic Advertising
PDU's Common Extended Header Format is manipulated.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix parent PDU's aux offset in aux ptr field

Fix Extended Advertising and Periodic Advertising's parent
PDU's aux ptr field to contain correct aux offset value to
its chain PDU.

Uninitialized pointer reference to aux ptr in the parent PDU
was used before the pointer reference was returned by the
function adding the aux ptr fields in the parent PDU.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
samples: Bluetooth: Add Extended Scanning support to Observer sample

Added implementation to Observer sample so that when
Extended Advertising feature is enabled, it can receive
Extended Advertising Reports, and hence receive long AD
data from peer Extended Advertising Broadcaster Multiple
sample.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

jori-nordic commented on samples/bluetooth/broadcaster_multiple/README.rst:
```diff
@@ -0,0 +1,23 @@
+.. _bluetooth-broadcaster-multiple-sample:
+
+Bluetooth: Broadcaster Multiple
```
I know it's "Broadcaster Multiple (advertising sets)" but I think "Multiple Broadcaster" reads better. what do you think

jori-nordic commented on samples/bluetooth/broadcaster_multiple/README.rst:
```diff
@@ -0,0 +1,23 @@
+********************
+
+This sample can be found under
+:zephyr_file:`samples/bluetooth/broadcast_multiple` in the Zephyr tree.
```
broadcaster

jori-nordic commented on samples/bluetooth/broadcaster_multiple/boards/nrf52_bsim.conf:
```diff
@@ -0,0 +1,18 @@
+
+# Zephyr Controller will use chain PDUs when AD data length > 191 bytes
+# - 31 bytes will use 22 bytes for the default name in this sample plus 9 bytes
+#   for manufacterer data
```
manufacturer

jori-nordic commented on samples/bluetooth/broadcaster_multiple/src/main.c:
```diff
@@ -0,0 +1,92 @@
+		}
+
+		/* Set extended advertising data */
+		err = bt_le_ext_adv_set_data(adv[index], ad, ARRAY_SIZE(ad),
```
so all the advertising sets contain the same data? could we show how to have different data per set instead?

jori-nordic reviewed (changes requested):
didn't review the controller part, sample seems fine, albeit a little light in demonstrated functionality.

rugeGerritsen commented on samples/bluetooth/observer/src/main.c:
```diff
@@ -41,13 +109,17 @@ void main(void)
-
 	printk("Bluetooth initialized\n");
 
+#if defined(CONFIG_BT_EXT_ADV)
+	bt_le_scan_cb_register(&scan_callbacks);
+	printk("Registered scan callbacks\n");
+#endif /* CONFIG_BT_EXT_ADV */
```
Why are we only registering scan callbacks when extended advertising is enabled?

Thalley replied:
I was thinking the same. We should retire the `device_found` callback and only use the `scan_callbacks` IMO.

Thalley replied:
> It is outside the scope of this PR to retire device_found.

It is fine to keep as is for me. 

I expect the reason why we have 2 scan callbacks is to avoid breaking a stable API, but might make sense to do at some point.

rugeGerritsen commented on samples/bluetooth/observer/src/main.c:
```diff
@@ -14,15 +14,83 @@
 
 	bt_addr_le_to_str(addr, addr_str, sizeof(addr_str));
-	printk("Device found: %s (RSSI %d)\n", addr_str, rssi);
+	printk("Device found: %s (RSSI %d) type %u, AD data len %u\n",
```
```suggestion
	printk("Device found: %s (RSSI %d), type %u, AD data len %u\n",
```

rugeGerritsen commented on samples/bluetooth/observer/src/main.c:
```diff
@@ -14,15 +14,83 @@
+};
+#endif /* CONFIG_BT_EXT_ADV */
+
 void main(void)
```
I would expect the sample documentation to be updated as the sample behavior has changed

rugeGerritsen commented on samples/bluetooth/observer/src/main.c:
```diff
@@ -14,15 +14,83 @@
+	bt_addr_le_to_str(info->addr, le_addr, sizeof(le_addr));
+	printk("[DEVICE]: %s, AD evt type %u, Tx Pwr: %i, RSSI %i "
+	       "Data status: %u, AD data len: %u Name: %s "
+	       "C:%u S:%u D:%u SR:%u E:%u Prim: %s, Secn: %s, "
```
What is `Secn` abbreviating?

rugeGerritsen commented on samples/bluetooth/observer/prj.conf:
```diff
@@ -1,2 +1,9 @@
+# CONFIG_BT_EXT_ADV=y
+# CONFIG_BT_EXT_SCAN_BUF_SIZE=1650
+
+# Zephyr Controller needs 16 buffers to generate Extended Advertising Report
```
```suggestion
# Zephyr Controller needs 16 buffers to be able to receive the full 1650 bytes of data
```

rugeGerritsen commented on samples/bluetooth/observer/boards/nrf52_bsim.conf:
```diff
@@ -0,0 +1,7 @@
+# Example Zephyr Controller setting to receive Extended Advertising Report
+
+# Set maximum scan data length for Extended Scanning
+# CONFIG_BT_CTLR_SCAN_DATA_LEN_MAX=1650
```
Is it needed to add a `nrf52_bsim.conf` when these configurations are already set in the `prj.conf`?

rugeGerritsen commented on samples/bluetooth/broadcaster_multiple/README.rst:
```diff
@@ -0,0 +1,23 @@
+This sample can be found under
+:zephyr_file:`samples/bluetooth/broadcaster_multiple` in the Zephyr tree.
+
+See :ref:`bluetooth samples section <bluetooth-samples>` for details.
```
I would expect more extensive documentation describing what the sample does, and how to validate it. See for example https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/samples/bluetooth/peripheral_hr_coded/README.html

rugeGerritsen commented on samples/bluetooth/broadcaster_multiple/boards/nrf52_bsim.conf:
```diff
@@ -0,0 +1,18 @@
+#   bytes for manufacturer data
+# CONFIG_BT_CTLR_ADV_DATA_LEN_MAX=192
+
+# Enabling PDU chaining
+# CONFIG_BT_CTLR_ADVANCED_FEATURES=y
+# CONFIG_BT_CTLR_ADV_AUX_PDU_BACK2BACK=y
```
Shouldn't the default configuration make it possible to perform chaining?

rugeGerritsen replied:
Can you deduct if chaining is required from `CONFIG_BT_CTLR_ADV_DATA_LEN_MAX`?

cvinayak replied:
I guess so, let me try if setting the values in Kconfig and update the PR, if I succeed.

rugeGerritsen commented on samples/bluetooth/broadcaster_multiple/src/main.c:
```diff
@@ -0,0 +1,103 @@
+static uint8_t mfg_data[BT_MFG_DATA_LEN] = { 0xFF, 0xFF, };
+
+static const struct bt_data ad[] = {
+	BT_DATA(BT_DATA_MANUFACTURER_DATA, mfg_data, sizeof(mfg_data)),
```
Does this really work when `sizeof(mfg_data) > UINT8_MAX`?

Thalley replied:
It would not, as the length field is 8-bit.

rugeGerritsen commented on samples/bluetooth/broadcaster_multiple/src/main.c:
```diff
@@ -0,0 +1,103 @@
+ * Bluetooth Core Specification define minimum of 31 bytes will be supported by
+ * all Controllers, can be a maximum of 1650 bytes when supported.
+ */
+#if defined(CONFIG_BT_LL_SW_SPLIT)
+#define BT_AD_DATA_LEN_MAX CONFIG_BT_CTLR_ADV_DATA_LEN_MAX
+#else
```
It would be nice if we could remove ZLL specific configurations from the sample

rugeGerritsen commented on samples/bluetooth/broadcaster_multiple/src/main.c:
```diff
@@ -0,0 +1,103 @@
+
+void main(void)
+{
+	struct bt_le_adv_param adv_param = {
```
Should we use `BT_LE_EXT_ADV_NCONN_NAME` instead?

rugeGerritsen commented on samples/bluetooth/broadcaster_multiple/prj.conf:
```diff
@@ -0,0 +1,7 @@
+CONFIG_BT_EXT_ADV=y
+CONFIG_BT_EXT_ADV_MAX_ADV_SET=2
+CONFIG_BT_DEVICE_NAME="Broadcaster Multiple"
+
```
Should we set `CONFIG_BT_CTLR_ADV_DATA_LEN_MAX` in the prj.conf?

Thalley replied:
I assume that this comment is no longer valid as we have `CONFIG_BT_CTLR_ADV_DATA_LEN_MAX=192` in this config file now

rugeGerritsen reviewed (changes requested):
I just looked at the updated samples. 

As the samples may be a starting point for customers trying out new functionality, we should ensure they are properly documented

Thalley commented on samples/bluetooth/broadcaster_multiple/src/main.c:
```diff
@@ -0,0 +1,103 @@
+		return;
+	}
+
+	index = CONFIG_BT_EXT_ADV_MAX_ADV_SET;
+	while (index--) {
```
For something like this, a `for` loop with `ARRAY_SIZE(adv)` might have been easier to understand (as you are iterating over the `adv` array here).

Thalley commented on samples/bluetooth/observer/src/main.c:
```diff
@@ -14,15 +14,83 @@
+	case BT_DATA_NAME_SHORTENED:
+	case BT_DATA_NAME_COMPLETE:
+		len = MIN(data->data_len, NAME_LEN - 1);
+		memcpy(name, data->data, len);
```
```suggestion
		(void)memcpy(name, data->data, len);
```

Thalley commented on samples/bluetooth/observer/src/main.c:
```diff
@@ -14,15 +14,83 @@
+static const char *phy2str(uint8_t phy)
+{
+	switch (phy) {
+	case 0: return "No packets";
```
```suggestion
	case BT_GAP_LE_PHY_NONE: return "No packets";
```

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -1385,8 +1385,8 @@ uint8_t ull_adv_aux_hdr_set_clear(struct ll_adv_set *adv,
 	struct pdu_adv_com_ext_adv *sec_com_hdr, *sec_com_hdr_prev;
-	struct pdu_adv_ext_hdr *pri_hdr, pri_hdr_prev;
-	struct pdu_adv_ext_hdr *sec_hdr, sec_hdr_prev;
+	struct pdu_adv_ext_hdr *hdr, pri_hdr, pri_hdr_prev;
```
With so many `hdr` variables in this function, just using `hdr` seems a bit too generic. 

Would it make sense to call it something like `ext_hdr`?

Thalley replied:
Ah, so `hdr` is used in multiple different contexts, and does not refer to a specific type of header? In that case I think it is OK to keep as is.

cvinayak replied:
yes, `hdr` is used in multiple context temporarily.

cvinayak committed:
samples: Bluetooth: Refactor observer for bsim test reuse

Refactor the observer sample so that it can be reused as
bsim test to verify Extended advertisement chaining.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
samples: Bluetooth: Refactor broadcaster multiple for bsim test reuse

Refactor the broadcaster multiple sample so that it can be
reused as bsim test to verify Extended advertisement
chaining.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
tests: Bluetooth: bsim: Verify multiple broadcaster and chain PDU

Added BabbleSim test to reuse the Broadcaster Multiple and
Oberver samples to verify multiple advertising sets and
transmission of Extended Advertising Chain PDU.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

jori-nordic commented on samples/bluetooth/observer/prj.conf:
```diff
@@ -1,2 +1,17 @@
 CONFIG_BT=y
 CONFIG_BT_OBSERVER=y
+
```
instead of doing this, could we follow `peripheral_hr` that uses an alternate `prj.conf` (`prj_minimal.conf`)?
And then let the user know (in the broadcaster sample) (s)he has to use `prj_extended.conf`.
I think it's better than commented out options.

cvinayak replied:
done

jori-nordic commented on tests/bluetooth/bsim_bt/bsim_test_adv_chain/tests_scripts/adv_chain.sh:
```diff
@@ -0,0 +1,38 @@
+# Copyright 2018 Oticon A/S
+# SPDX-License-Identifier: Apache-2.0
+
+# Multiple connection between two devices with multiple peripheral identity
```
is that correct?

cvinayak replied:
fixed

cvinayak commented:
@Thalley please re-review

cvinayak committed:
Bluetooth: Controller: Enable back-to-back PDU chaining

When AD data length of greater than 191 bytes support is
enabled then default to using the back-to-back PDU chaining
feature in the Controller.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on samples/bluetooth/broadcaster_multiple/src/broadcaster_multiple.c:
```diff
@@ -0,0 +1,113 @@
+ * Bluetooth Core Specification define minimum of 31 bytes will be supported by
+ * all Controllers, can be a maximum of 1650 bytes when supported.
+ */
+#if defined(CONFIG_BT_CTLR_ADV_DATA_LEN_MAX)
+#define BT_AD_DATA_LEN_MAX CONFIG_BT_CTLR_ADV_DATA_LEN_MAX
+#else
+#define BT_AD_DATA_LEN_MAX 31U
+#endif
```
Since `CONFIG_BT_CTLR_ADV_DATA_LEN_MAX` is only guarded by `CONFIG_BT_CTLR_ADV_EXT`, and since this sample requires extended advertising, I think this guard here can be removed, right?

Thalley replied:
Ah, I see. 

I guess the host should actually read the maximum supported advertising length with `BT_HCI_OP_LE_READ_MAX_ADV_DATA_LEN` in those cases, but it isn't implemented in the Zephyr host. 

It seems that I even created an issue for this a while ago: https://github.com/zephyrproject-rtos/zephyr/issues/45230

Thalley commented on samples/bluetooth/observer/prj_extended.conf:
```diff
@@ -0,0 +1,17 @@
+CONFIG_BT_EXT_ADV=y
+CONFIG_BT_EXT_SCAN_BUF_SIZE=1650
+
+# Zephyr Bluetooth LE Controller needs 16 event buffers to generate Extended
+# Advertising Report for receiving the complete 1650 bytes of data
+CONFIG_BT_BUF_EVT_RX_COUNT=16
```
Is this really true? 

On a single core board, I would expect it to be something like

1. Controller generates event
2. Host handles event (in same thread)
3. Host releases memory for event
4. Go to 1. 

And for the dual core boards (e.g. NRF5340) the host should be able to handle the events as they come, right? 

If this is indeed the case (i.e. this value needs to be 16 to handle 1650 octets of data), then perhaps the default value should be changed? 

We have `CONFIG_BT_EXT_SCAN_BUF_SIZE` which determines the size of the events we can handle in the host. Since each report may be up to 229 octets, the number of events we need to handle should thus be `CONFIG_BT_EXT_SCAN_BUF_SIZE / 229`. 

So in case that we cannot handle each event as they come, does `CONFIG_BT_BUF_EVT_RX_COUNT` really need to be larger than 8? The default is `10` unless `CONFIG_BT_RECV_BLOCKING=y (and in which case, that should also work, right?).

cvinayak replied:
> Is this really true?

This is true. In the *best*  case,  ceil(1650 / 229) = 8 HCI reports are generated in one go by the Controller's Rx thread and while the host is recombining these 8 reports, to keep the continuity (not block the Rx thread) of receiving the reports at higher rate, 8 additional buffer for the report are allocated to realize double buffering. I.e. while 8 reports are being recombined by host, 8 new reports can be populated by the Controller's Rx Thread without deadlocking the same Rx thread that is recombining.

> And for the dual core boards (e.g. NRF5340) the host should be able to handle the events as they come, right?

Yes, for the host (application core) only sample `CONFIG_BT_BUF_EVT_RX_COUNT=1` should suffice to receive the reports from the transport, but on the Controller (network core) `CONFIG_BT_BUF_EVT_RX_COUNT=16` is required to buffer the reports until the reports are transmitted out over the transport (RPMSG/UART/SPI etc.)

> If this is indeed the case (i.e. this value needs to be 16 to handle 1650 octets of data), then perhaps the default value should be changed?

In a *worst* case a peer can send chain PDUs of 1 byte AD data, Controller does not support recombining the PDUs into report. In a typical usecase only one AUX_ADV_IND PDU is used (upto 191 byte AD data) with no chaining, under such cases the Kconfig default `CONFIG_BT_BUF_EVT_RX_COUNT=3` (under `CONFIG_BT_RECV_BLOCKING=y` is sufficient (only 1 buffer is really needed).

> So in case that we cannot handle each event as they come, does CONFIG_BT_BUF_EVT_RX_COUNT really need to be larger than 8? The default is 10 unless `CONFIG_BT_RECV_BLOCKING=y (and in which case, that should also work, right?).

As explained in first reply, in order to keep the Controller Rx Thread from blocking on allocating report buffers while the 8 reports are being recombined, 8 additional report buffers are required which correspond to the new report that is being prepared from the PDUs that have arrived on `CONFIG_BT_CTLR_RX_BUFFERS=9` (9 because Controller needs to receive the ADV_EXT_IND + AUX_ADV_IND + 7 AUX_CHAIN_IND PDUs).

Thalley replied:
I still think I am missing some understanding here. 

In the case of `CONFIG_BT_RECV_BLOCKING=y`, why does the controller need to generate multiple events in one go? Wouldn't the controller just create a single event per advertisement it receives over air, and forward that to the host (which shouldn't require more than 1 event as it is done in the same thread). 

In the case of `CONFIG_BT_RECV_BLOCKING=n` (with the default event count of 10) the host should be able handle the events as they come, right? You even say that yourself: 

> Yes, for the host (application core) only sample CONFIG_BT_BUF_EVT_RX_COUNT=1 should suffice to receive the reports from the transport, but on the Controller (network core) CONFIG_BT_BUF_EVT_RX_COUNT=16 is required to buffer the reports until the reports are transmitted out over the transport (RPMSG/UART/SPI etc.)

Is this configuration file intended to be used on the network core too? Otherwise the increase in `CONFIG_BT_BUF_EVT_RX_COUNT` here doesn't help the network core.

cvinayak replied:
> why does the controller need to generate multiple events in one go?

The controller implementation (choice) buffers received PDU until it can decide if the last report would have a "data complete", or "data incomplete, no more to come" as data status, hence all the reports are generated in one go at the end of the reception of the chain.

Also, a single PDU that can be > 229 bytes of AD data will fragment into two advertising report, `CONFIG_BT_BUF_EVT_RX_COUNT=1` will deadlock the Rx thread.

In the case of periodic advertising reports, the implementation is different where reports are generated as the PDU arrives. This is different here as we had to interleave the IQ reports when direction finding was supported and used. Periodic Advertising report generated has in fact gotten complicated, as a additional/extra dummy report is generated with "data complete" or "data incomplete, no more to come" based on whether there will be more PDUs or the peer stops anymore PDU transmission (or fail to receive the subsequent PDUs).

Thalley replied:
> The controller implementation (choice) buffers received PDU until it can decide if the last report would have a "data complete", or "data incomplete, no more to come" as data status, hence all the reports are generated in one go at the end of the reception of the chain.

Ah, that explains it (missed that part). 
So the controller will keep PDUs until all is received? I assume that it then aborts it if the number of PDUs (e.g. 100 PDUs of size 1 for a total of 100 octets) is larger than `CONFIG_BT_BUF_EVT_RX_COUNT`? 

Does it also handle the case where the advertising data is fragmented into e.g. 5 PDUs, P1, P2, ..., P5, and where e.g. P5 is lost, e.g.
-> P1
-> P2
-> P3
-> P4 
-/> P5 (lost)
-> P1 (should abort previously received P1 - P4 PDUs
?

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -1146,11 +1140,9 @@ uint8_t ll_adv_aux_sr_data_set(uint8_t handle, uint8_t op, uint8_t frag_pref,
-
 		/* Check AdvData overflow */
-		if ((sec_len + len) > PDU_AC_PAYLOAD_SIZE_MAX) {
+		if ((sec_len + ad_len_overflow + len) >
+		    PDU_AC_PAYLOAD_SIZE_MAX) {
```
Any risk of overflow here when comparing?

cvinayak replied:
`sec_len` is `uint16_t`; sum of `sec_len` (<256) , `ad_len_overflow` (<256) and `len` (<256) does not exceed `UINT16_MAX`

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_adv_chain/src/main.c:
```diff
@@ -0,0 +1,208 @@
+		return;
+	}
+
+	for (int i = 0; i < sid_count; i++) {
```
Suggest to use same type for iterator and the value it's compared against
```suggestion
	for (uint8_t i = 0; i < sid_count; i++) {
```

cvinayak replied:
Done.

rugeGerritsen commented on samples/bluetooth/broadcaster_multiple/prj.conf:
```diff
@@ -0,0 +1,23 @@
+
+CONFIG_BT_DEBUG_LOG=y
+
+# Zephyr Bluetooth LE Controller will need to use chain PDUs when AD data
+# length > 191 bytes
+# - 31 bytes will use 22 bytes for the default name in this sample plus 9 bytes
+#   for manufacturer data
+# - 191 bytes will use 22 bytes for the default name in this sample plus 169
+#   bytes for manufacturer data
+# - 277 bytes will use 22 bytes for the default name in this sample plus 255
+#   bytes for manufacturer data
+# CONFIG_BT_CTLR_ADV_DATA_LEN_MAX=192
+
+# Increase Advertising PDU buffers to number of advertising sets times the
+# number of chain PDUs per advertising set when using Zephyr Bluetooth LE
+# Controller
+# CONFIG_BT_CTLR_ADVANCED_FEATURES=y
+# CONFIG_BT_CTLR_ADV_DATA_BUF_MAX=2
```
Do we need all this commented out?

cvinayak replied:
Yes, so that host only sample like for qemu_x86 etc. can build in CI. These Kconfigs are covered by the added bsim test in this PR, and hence do not need to  explicitly be build tested by twister here for host+controller combined sample.

rugeGerritsen commented on samples/bluetooth/broadcaster_multiple/src/broadcaster_multiple.c:
```diff
@@ -0,0 +1,113 @@
+
+static uint8_t mfg_data[BT_MFG_DATA_LEN] = { 0xFF, 0xFF, };
+
+static const struct bt_data ad[] = {
+	BT_DATA(BT_DATA_MANUFACTURER_DATA, mfg_data, sizeof(mfg_data)),
+};
```
Did you consider adding more entries to this list to demonstrate how to send more than 255 bytes of data?

cvinayak replied:
I did manually test it by added more `BT_DATA_MANUFACTURER_DATA` entries, but not updated here in the sample, as the calculation of the subsequent size of the array and conditional compilation of additional entries here is complicated when it is to be derived from the `CONFIG_BT_CTLR_ADV_DATA_LEN_MAX` value.

rugeGerritsen reviewed (approved):
Checked the samples only
